pop-api/integration-tests/src/local_fungibles.rs

```diff
@@ -625,3 +352,269 @@ fn token_metadata_works() {
 // 		assert!(!result.did_revert(), "Contract reverted!");
 // 	});
 // }
+
+fn decoded<T: Decode>(result: ExecReturnValue) -> Result<T, ExecReturnValue> {
+	<T>::decode(&mut &result.data[1..]).map_err(|_| result)
+}
+
+// Call total_supply contract message.
+fn total_supply(addr: AccountId32, asset_id: AssetId) -> Result<Balance, Error> {
+	let function = function_selector("total_supply");
+	let params = [function, asset_id.encode()].concat();
+	let result = bare_call(addr, params, 0).expect("should work");
+	match decoded::<Result<Balance, Error>>(result) {
+		Ok(x) => x,
+		Err(result) => panic!("Contract reverted: {:?}", result),
+	}
+}
+
+// Call balance_of contract message.
+fn balance_of(addr: AccountId32, asset_id: AssetId, owner: AccountId32) -> Result<Balance, Error> {
+	let function = function_selector("balance_of");
+	let params = [function, asset_id.encode(), owner.encode()].concat();
+	let result = bare_call(addr, params, 0).expect("should work");
+	match decoded::<Result<Balance, Error>>(result) {
```

Comment:
These lines can be cleaner if we use `map_err`
```rs
decoded::<Result<Balance, Error>>(result).map_err(|_| panic!("Contract reverted: {:?}", result))
```

Reply:
Another thing is this match logic is reused for all the other methods, do you think we should have another shared method called `contract_exec_decoded` or something else that can be reused for all the other methods?

Reply:
Regarding your first comment; the code will then return a `Result<Result<_, _>>` which we don't want. 

Second comment; the problem with that is when the panic occurs it is not clear from which test it is coming and which check within that test. As a result, we have more duplicated code but know immediately which check is failing which improves development.

Reply:
To your first comment, I will refactor it to:
```
decoded::<Result<(), Error>>(result).unwrap_or_else(|_| panic!("Contract reverted: {:?}", result))
```